Added SelectorDocument and SelectorValue.

GooDRomka opened the pull request:
@mxsnq please review new feature with selectors for documents and constants.

mxsnq commented on extension/scripts/Selector/SelectorDocument.js:
```diff
@@ -0,0 +1,92 @@
+    },
+
+    getFeatures: function () {
+        return ['selector','multiple', 'delay']
```
+downloadDocument, stringReplacement

mxsnq replied:
+ stringReplacement

mxsnq replied:
+ stringReplacement

mxsnq commented on extension/scripts/Selector/SelectorDocument.js:
```diff
@@ -0,0 +1,92 @@
+            dfd.resolve(result);
+        });
+
+        return dfd.promise();
```
Try to avoid copy-pasting big chunks of code from one module (SelectorLink) to another. This makes the code harder to maintain. Break repetitive logic down, move to separate units and call them as you need.

GooDRomka commented:
- SelectorValue and ConstantValue.
Reverted changes to SelectorValue and implemented ConstantValue selector.

- SelectorImage and SelectorDocument
We added stringReplacemant to SelectorDocument.
Now we integrated a function getFileFilename in Selector.js methods and changed  (downloadFileAsBlob, downloadImageBas64) to new function downloadFileAsBase64 in Selector.js. 
Now downloadFileAsBase64 return blob and also filename with using url

Yatskov commented on docs/Selectors/Value selector.md:
```diff
@@ -0,0 +1,6 @@
+# Value selector
+Value selector is input constant wich will be added to parent selector, like column with name selector "id" and contain "input constant"
```
This description is for Constant Value. We should add it in separate file and change this description for Selector Value description.

Yatskov commented on extension/devtools/devtools_scraper_panel.html:
```diff
@@ -22,7 +22,9 @@
     <script src="../scripts/Selector/SelectorGroup.js"></script>
     <script src="../scripts/Selector/SelectorText.js"></script>
     <script src="../scripts/Selector/SelectorValue.js"></script>
```
SelectorValue is name that hard to
to distinguish from ConstantValue. I suggest to rename it to InputValue.

Yatskov commented on extension/devtools/views/SelectorEdit.html:
```diff
@@ -172,13 +184,22 @@
+            <input type="text" class="form-control" name="constantValue" id="constantValue" placeholder="Text constant to insert" value="{{selector.constantValue}}">
+        </div>
+    </div>
+
```
I suggest we can use the same feature insertValue for both InputValue and ConstantValue.

Yatskov commented on extension/scripts/Selector/ConstantValue.js:
```diff
@@ -0,0 +1,51 @@
+    },
+
+    canHaveChildSelectors: function () {
+        return true;
```
ConstantValue could not have child selectors.

Yatskov commented on extension/scripts/Selector/ConstantValue.js:
```diff
@@ -0,0 +1,51 @@
+    },
+
+    canHaveLocalChildSelectors: function () {
+        return true;
```
ConstantValue could not have local child selectors.

Yatskov commented on extension/scripts/Selector/ConstantValue.js:
```diff
@@ -0,0 +1,51 @@
+        var result = [];
+        $(elements).each(function (k, element) {
+            $(element).val(this.constantValue);
+        }.bind(this));
```
How can we get data elements if we do not have selector in ConstantValue as a feature?

Yatskov commented on extension/scripts/Selector/SelectorDocument.js:
```diff
@@ -0,0 +1,110 @@
+                    }
+
+                    href = href.replace(replace, replacement);
+                }
```
I'd suggest to move string replacement in SelectorLink, SelectorDocument and SelectorImage to Selector class.

Yatskov commented on extension/scripts/Selector/SelectorDocument.js:
```diff
@@ -0,0 +1,110 @@
+                data[this.id] = $(element).text();
+                // data._followSelectorId = this.id;
+                data[this.id + '-href'] = href;
+                // data._follow = href;
```
Please, delete unnecessary comments.

Yatskov commented on extension/scripts/Selector/SelectorDocument.js:
```diff
@@ -0,0 +1,110 @@
+
+    // getItemCSSSelector: function() {
+    // 	return ["img", "div"];
+    // }
```
Unnecessary comments?

Yatskov commented on extension/scripts/Selector.js:
```diff
@@ -239,6 +239,50 @@ var Selector = (function () {
+            if(filename.length > 130) {
+                filename = filename.substr(0, 130);
+            }
+            return filename;
```
We need a better error handling here.

GooDRomka commented:
@Yatskov please review new chenges.
I hope this is the last try.

Yatskov commented on extension/scripts/Selector.js:
```diff
@@ -239,6 +258,50 @@ var Selector = (function () {
+        },
+
+
+        getFileFilename: function(url) {
```
Не очень понятное имя функции. Предложил бы getFilenameFromUrl

Yatskov commented on extension/scripts/Selector.js:
```diff
@@ -215,6 +215,25 @@ var Selector = (function () {
             }
         },
 
+        StringReplacement: function(url, stringReplacement){
```
Имена методов принято начинать со строчного символа, предлагаю изменить в stringReplacement.

GooDRomka replied:
stringReplacement используется для названия фичи селектора. Предлагаю назвать stringReplace

Yatskov commented on extension/devtools/views/SelectorEdit.html:
```diff
@@ -172,7 +184,7 @@
+        <label for="insertValue" class="col-sm-2 control-label">Value</label>
 
         <div class="col-sm-8">
             <input type="text" class="form-control" name="insertValue" id="insertValue" placeholder="Text value to insert" value="{{selector.insertValue}}">
```
I'd suggest to rename insertValue to value.
